docs/man/mctally.1
.Dd January 2, 2024
.Dt mctally 1
.Os mcplex.net
.Sh NAME
.Nm mctally
.Nd request list of installed packages matching a regular expression
.Sh SYNOPSIS
.Nm
.Op Fl h Ar host(s)
.Op Fl a
.Op Fl l
.Op Fl p Ar port
.Op Fl t
.Op Fl u
.Op Fl v
.Op packages_regex
.Sh DESCRIPTION
.Nm
requests installed package information from one or more hosts via
.Xr mctallyd 8 .
It can also request load averages, active logins, uptime and uname
information.
.Pp
The following command line options are available:
.Pp
.Bl -tag -width indent
.It Fl h Ar host(s)
Specify the
.Xr mctallyd 8
host(s) to be queried.  If not specified, localhost is queried.
.It Fl a
Request load averages.
.It Fl l
Request active logins.
.It Fl p Ar port
Specify the TCP port where
.Xr mctallyd 8
is listening.
.It Fl t
Request uptime.
.It Fl u
Request uname information.
.It Fl v
Display the version of mctally.
.It packages_regex
Request a list of installed packages whose name matches \fIpackages_regex\fR.
\fIpackages_regex\fR is an ECMAScript regular expression.
.El
.Sh FILES
.Nm
uses libCredence for encryption and authentication.  Hence it
requires credence key files.
.Bl -tag -width indent
.It Pa ~/.credence/id_ed25519
The user's credence private key file, created with
.Xr credence 1 .  This file should be owned by the user and have
permissions 0600.  It must contain the private part of an Ed25519
RSA key pair.
.It Pa ~/.credence/id_ed25519.pub
The user's credence public key file, created with
.Xr credence 1 .  This file should be owned by the user and have
permissions 0600.  It must contain the public part of an Ed25519 key pair.
.It Pa ~/.credence/known_keys
The user's credence known services file.  This file must contain the public
keys of
.Xr mctallyd 8
from the
.Xr mctallyd 8
hosts being queried (one per line).  It is used to authenticate instances
of
.Xr mctallyd 8 .
.El
.Sh SEE ALSO
.Xr mctallyd 8 ,
.Xr credence 1
.Sh COPYRIGHT
Copyright (c) 2024 Daniel W. McRobb
.Mt dwm@mcplex.net
.Pp



ChangeLog
2024-07-28  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as mctally-0.1.12

2024-07-28  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* docs/man/mctally.1: Document the -v command line option.
	* apps/mctally/mctally.cc: Added -v command line option to display
	the version of mctally.

2024-07-26  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as mctally-0.1.11

	* docs/man: Added manpages for mctally and mctallyd.

	* apps/mctally/mctally.cc: Add usage statement.  Allow port number
	to be specified on command line.  Print error on cerr if we fail
	to connect, authenticate, send or receive.

	* tagged as mctally-0.1.10

	* classes/include/DwmMcTallyInstalledPackages.hh
	* classes/include/DwmMcTallyLoadAvg.hh
	* classes/include/DwmMcTallyLogins.hh
	* classes/include/DwmMcTallyRequest.hh
	* classes/include/DwmMcTallyResponse.hh
	* classes/include/DwmMcTallyUname.hh

	Don't need to inherit from StreamIOCapable with newer libDwm.

2024-06-01  Daniel W. McRobb  <dwm@kiva.mcplex.net>

	* tagged as mctally-0.1.9

	* Makefile, apps/mctally/Makefile, apps/mctallycgi/Makefile:
	apps/mctallyd/Makefile, classes/tests/Makefile: Fix up
	'distclean' target.

2024-01-27  Daniel McRobb  <dwm@spark.mcplex.net>

	* tagged as mctally-0.1.8

	* apps/mctallycgi/mctallycgi.cc: Added toto to list of hosts.
	Added Mcweather to list of apps.

2024-01-06  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as mctally-0.1.7

	* Added uptime request/response.

2024-01-05  Daniel W. McRobb  <dwm@www.mcplex.net>

	* apps/mctallycgi/mctallycgi.cc: Hardcode key directory path.
	Don't forget to emit Content-type.

	* tagged as mctally-0.1.6

	* Added apps/mctalycgi for my own purposes.

2024-01-04  Daniel McRobb  <dwm@spark.mcplex.net>

	* apps/mctally/mctally.cc: Refactor very ugly code (reduced
	lines of code by about 50%).  Use a thread per host to avoid
	waiting for each one serially.

	* classes/src/DwmMcTallyUtils.cc: Catch some regex exceptions.

	* classes/src/DwmMcTallyUtils.cc: include MacPorts in search
	for local packages on macOS.  Prefix them with 'org.macports.'.

	* apps/mctally/mctally.cc: Allow multiple hosts (separated by
	commas) to be specified in each -h command line optarg.

2024-01-03  Daniel W. McRobb  <dwm@kiva.mcplex.net>

	* tagged as mctally-0.1.5

2024-01-03  dwm  <dwm@nomnom.mcplex.net>

	* apps/mctally/mctally.cc: Need <iomanip>

	* classes/src/DwmMcTallyLogins.cc:  Need <mutex>

2024-01-03  Daniel McRobb  <dwm@spark.mcplex.net>

	* tagged as mctally-0.1.4

	* A bunch of work toward a usable protocol.  Now support fetching
	of active logins, load averages, uname and installed packages.

2024-01-02  Daniel McRobb  <dwm@spark.mcplex.net>

	* classes/include/DwmMcTallyLogins.hh
	* classes/src/DwmMcTallyLogins.cc
	* classes/src/Makefile
	* classes/tests/.gitignore
	* classes/tests/Makefile
	* classes/tests/TestTallyLogins.cc
	Added Logins class and unit tests.

	* classes/include/DwmMcTallyInstalledPackages.hh
	* classes/src/DwmMcTallyInstalledPackages.cc
	* classes/src/Makefile
	* classes/tests/.gitignore
	* classes/tests/Makefile
	* classes/tests/TestTallyInstalledPackages.cc
	Added InstalledPackages class and unit tests.  Working my way
	toward encapsulating a real protocol.

	* apps/mctally/mctally.cc: Allow multiple hosts (-h can be repeated
	on command line).

2024-01-01  Daniel McRobb  <dwm@spark.mcplex.net>

	* classes/include/DwmMcTallyLoadAvg.hh,
	classes/src/DwmMcTallyLoadAvg.cc, classes/tests/TestTallyLoadAvg.cc,
	classes/tests/Makefile:
	Add LoadAvg class and unit tests.

2023-12-31  Daniel McRobb  <dwm@spark.mcplex.net>

	* classes/tests/TestTallyUname.cc: added unit tests for JSON
	functionality.

	* classes/include/DwmMcTallyUname.hh, classes/src/DwmMcTallyUname.cc:
	Added ToJson() and FromJson() members.

2023-12-30  Daniel W. McRobb  <dwm@nomnom.mcplex.net>

	* tagged as mctally-0.1.3

	* getvers.sh: We're mctally, not Mcweather.

2023-12-30  Daniel W. McRobb  <dwm@nomnom.mcplex.net>

	* tagged as mctally-0.1.2

2023-12-30  Daniel McRobb  <dwm@thrip.mcplex.net>

	* I don't have a doc directory yet, so don't try to clean it with
	'distclean' target.

	* tagged as mctally-0.1.1

2023-12-30  Daniel W. McRobb  <dwm@kiva.mcplex.net>

	* classes/tests/TestTallyUtils.cc: turn into actual unit tests.

2023-12-30  Daniel McRobb  <dwm@spark.mcplex.net>

	* configure, configure.ac, dwm.m4: Tweak message from
	DWM_REQUIRES_DWMPKG macro.

2023-12-29  Daniel McRobb  <dwm@spark.mcplex.net>

	* tagged as mctally-0.1.0

	* configure, configure.ac, dwm.m4: Add and use DWM_REQUIRES_DWMPKG
	macro.

2023-12-03  Daniel McRobb  <dwm@spark.mcplex.net>

	* Added ability to fetch uname information.

	* configure, configure.ac: Now generate apps/mctallyd/mctallyd.cc
	just so we can set default config file path according to @prefix@.
	Generate etc/macos/scripts/preinstall and
	etc/macos/scripts/postinstall.
	Generate etc/net.mcplex.mctallyd.plist

	* apps/mctallyd/mctallyd.c.in: Added -d (don't daemonize) command
	line option to mctallyd	so we can run under launchd on macOS.

	* etc/macos/scripts/preinstall.in:
	* etc/macos/scripts/postinstall.in:
	Preinstall and postinstall scripts for macOS.  Do what we need to
	do with launchdctl and set up installing user to authenticate to
	mctallyd if they have an existing credence key.

	* etc/net.mcplex.mctallyd.plist.in:
	launchd daemon plist for mctallyd.



apps/mctally/mctally.cc
//===========================================================================
// @(#) $DwmPath$
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2023, 2024
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  \file mctally.cc
//!  \author Daniel W. McRobb
//!  \brief client program to show installed applications, uptime, uname,
//!    load averages or active logins
//---------------------------------------------------------------------------

extern "C" {
  #include <unistd.h>
}

#include <iomanip>
#include <iostream>
#include <regex>
#include <thread>

#include "DwmSysLogger.hh"
#include "DwmCredencePeer.hh"
#include "DwmMcTallyResponse.hh"
#include "DwmMcTallyUname.hh"
#include "DwmMcTallyUtils.hh"
#include "DwmMcTallyVersion.hh"

using namespace std;
using namespace Dwm;

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static vector<string> SplitArg(const string & arg)
{
  vector<string>  rc;
  size_t   startIndex = 0;
  size_t   foundIndex;
  while ((foundIndex = arg.find_first_of(',',startIndex)) != string::npos) {
    string  argpiece = arg.substr(startIndex, foundIndex - startIndex);
    if (! argpiece.empty()) {
      rc.push_back(arg.substr(startIndex, foundIndex - startIndex));
    }
    startIndex = foundIndex + 1;
  }
  if (startIndex < arg.size()) {
    string  argpiece = arg.substr(startIndex);
    if (! argpiece.empty()) {
      rc.push_back(arg.substr(startIndex));
    }
  }
  return rc;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool GetPeer(const string & host, uint16_t port, Credence::Peer & peer)
{
  bool  rc = false;
  if (peer.Connect(host, port)) {
    Credence::KeyStash   keyStash;
    Credence::KnownKeys  knownKeys;
    if (peer.Authenticate(keyStash, knownKeys)) {
      rc = true;
    }
    else {
      peer.Disconnect();
      cerr << "Failed to authenticate with " << host << ':' << port << '\n';
    }
  }
  else {
    cerr << "Failed to connect to " << host << ':' << port << '\n';
  }
  return rc;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintInstalledPackages(const string & host,
                                   const McTally::Response & resp)
{
  if (std::holds_alternative<McTally::InstalledPackages>(resp.Data())) {
    const auto & installedPkgs =
      std::get<McTally::InstalledPackages>(resp.Data());
    cout << host;
    if (installedPkgs.Pkgs().empty()) {
      cout << ": no";
    }
    cout << " installed packages matching '" << resp.Req().Selector() << "'";
    if (installedPkgs.Pkgs().empty()) {
      cout << '\n';
    }
    else {
      cout << ":\n";
      for (const auto & pkg : installedPkgs.Pkgs()) {
        cout << "  " << pkg.first << ' ' << pkg.second << '\n';
      }
    }
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintUname(const string & host, const McTally::Response & resp)
{
  if (std::holds_alternative<McTally::Uname>(resp.Data())) {
    const auto & un = std::get<McTally::Uname>(resp.Data());
    cout << host << " uname:\n"
         << "  sysname:    " << un.SysName() << '\n'
         << "  nodename:   " << un.NodeName() << '\n'
         << "  release:    " << un.Release() << '\n'
         << "  version:    " << un.Version() << '\n'
         << "  machine:    " << un.Machine() << '\n'
         << "  prettyname: " << un.PrettyName() << '\n';
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintLoginEntries(const string & host,
                              const McTally::Logins & logins)
{
  ostringstream  oss;
  if (! logins.Entries().empty()) {
    oss << setiosflags(ios::left)
        << "  " << setw(10) << "USER" << ' ' << setw(8) << "TTY" << ' '
        << setw(16) << "LOGIN_TIME" << ' ' << setw(16) << "IDLE_TIME" << ' '
        << "FROM\n";
    for (const auto & entry : logins.Entries()) {
      oss << "  " << setw(10) << entry.User() << ' '
          << setw(8) << entry.Tty() << ' '
          << setw(16) << entry.LoginTimeString() << ' '
          << setw(16) << entry.IdleTimeString() << ' '
          << entry.FromHost() << '\n';
    }
    cout << host << " active logins:\n" << oss.str();
  }
  else {
    cout << host << ": no active logins\n";
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintLogins(const string & host, const McTally::Response & resp)
{
  if (std::holds_alternative<McTally::Logins>(resp.Data())) {
    const auto & logins = std::get<McTally::Logins>(resp.Data());
    PrintLoginEntries(host, logins);
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintLoadAverages(const string & host,
                              const McTally::Response & resp)
{
  if (std::holds_alternative<McTally::LoadAvg>(resp.Data())) {
    const auto & loadAvgs = std::get<McTally::LoadAvg>(resp.Data());
    cout << setiosflags(ios::left)
         << setw(40) << host + " load averages: "
         << setprecision(4)
         << setw(6) << fixed << loadAvgs.Avg1Min() << ' '
         << setw(6) << fixed << loadAvgs.Avg5Min() << ' '
         << setw(6) << fixed << loadAvgs.Avg15Min() << '\n';
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintUptime(const string & host, const McTally::Response & resp)
{
  if (std::holds_alternative<uint64_t>(resp.Data())) {
    auto  uptime = std::get<uint64_t>(resp.Data());
    cout << host << " uptime: "
         << McTally::Utils::TimeIntervalString(uptime) << '\n';
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintResponse(const string & host, const McTally::Response & resp)
{
  switch (resp.Req().ReqEnum()) {
    case McTally::e_installedPackages:
      PrintInstalledPackages(host, resp);
      break;
    case McTally::e_uname:
      PrintUname(host, resp);
      break;
    case McTally::e_loadAverages:
      PrintLoadAverages(host, resp);
      break;
    case McTally::e_logins:
      PrintLogins(host, resp);
      break;
    case McTally::e_uptime:
      PrintUptime(host, resp);
      break;
    default:
      break;
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintResponses(const string & host,
                           const vector<McTally::Response> & responses)
{
  static mutex  mtx;
  lock_guard<mutex>  lock(mtx);
  for (const auto & resp : responses) {
    PrintResponse(host, resp);
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool SendReceive(Credence::Peer & peer,
                        const vector<McTally::Request> & requests,
                        vector<McTally::Response> & responses)
{
  bool  rc = false;
  responses.clear();
  for (const auto & req : requests) {
    if (peer.Send(req)) {
      McTally::Response  resp;
      if (peer.Receive(resp)) {
        if (resp.Req().ReqEnum() == req.ReqEnum()) {
          responses.push_back(resp);
        }
      }
      else {
        cerr << "Failed to receive response from " << peer.EndPointString()
             << '\n';
      }
    }
    else {
      cerr << "Failed to send request to " << peer.EndPointString() << '\n';
    }
  }
          
  return (requests.size() == responses.size());
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PeerThread(string host, uint16_t port,
                       const vector<McTally::Request> & requests)
{
  Credence::Peer  peer;
  if (GetPeer(host, port, peer)) {
    vector<McTally::Response>  responses;
    if (SendReceive(peer, requests, responses)) {
      PrintResponses(host, responses);
    }
    peer.Disconnect();
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void Usage(const char *argv0)
{
  cerr << "usage: " << argv0
       << " [-h host(s)] [-p port] [-a] [-l] [-t] [-u] [v] [packages_regex]\n"
       << "  -a: show load averages\n"
       << "  -l: show active logins\n"
       << "  -p port: connect to the given port (default 2125)\n"
       << "  -t: show uptime\n"
       << "  -u: show uname\n"
       << "  -v: show version\n"
       << "  -h hosts: specify hosts to query (comma-separated list)\n"
       << "  packages_regex: regular expression to match installed packages\n";
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
int main(int argc, char *argv[])
{
  extern int                optind;
  int                       optChar;
  vector<string>            hosts;
  uint16_t                  port = 2125;
  vector<McTally::Request>  requests;
  
  Dwm::SysLogger::Open("mctally", LOG_PERROR, LOG_USER);
  Dwm::SysLogger::MinimumPriority(LOG_ERR);
  Dwm::SysLogger::ShowFileLocation(true);

  while ((optChar = getopt(argc, argv, "h:p:ulatv")) != -1) {
    switch (optChar) {
      case 'a':
        requests.push_back(McTally::Request(McTally::e_loadAverages));
        break;
      case 'h':
        {
          vector<string>  newhosts = SplitArg(optarg);
          for (const auto & newhost : newhosts) {
            hosts.push_back(newhost);
          }
        }
        break;
      case 'p':
        port = std::stoul(optarg);
        break;
      case 'l':
        requests.push_back(McTally::Request(McTally::e_logins));
        break;
      case 't':
        requests.push_back(McTally::Request(McTally::e_uptime));
        break;
      case 'u':
        requests.push_back(McTally::Request(McTally::e_uname));
        break;
      case 'v':
        cout << Dwm::McTally::Version.Version() << '\n';
        return 0;
        break;
      default:
        Usage(argv[0]);
        return 1;
        break;
    }
  }

  if (hosts.empty()) {
    hosts.push_back("localhost");
  }

  vector<string>  regExps;
  for (int n = optind; n < argc; ++n) {
    try {
      regex  rgx(argv[n]);
      requests.push_back(McTally::Request(McTally::PackageSelector(argv[n])));
    }
    catch (...) {
      cerr << "Bad regular expression '" << argv[n] << "'\n";
      return 1;
    }
  }

  if (requests.empty()) {
    Usage(argv[0]);
    return 1;
  }
  
  vector<std::thread>  peerThreads;
  for (const auto & host : hosts) {
    peerThreads.emplace_back(thread(PeerThread, host, port,
                                    std::ref(requests)));
  }
  for (auto & peerThread : peerThreads) {
    peerThread.join();
  }
  
  return 0;
}

